src/Model.php

```diff
@@ -1178,12 +1178,14 @@ public function reload()
      */
     public function duplicate($newId = null)
     {
-        $this->setId(null);
+        $this->entityId = null;
```

Comment:
Entity approach must satisty several axioms. One of them is to never allow to assing different ID to the same instance.

So I think we should remove duplicate completely or clone before at the beginning of this method and return the cloned model then.

Reply:
Removal of method is not good option as functionality is useful.
Cloning model and returning it is a good option but `$entityId` is private. how to clear it best?

Reply:
I came up with following code but it does not work as even if model unloaded `$entityId` value remains saved. I think this is not the correct behavior

```
    public function duplicate($newId = null): self
    {
        $duplicate = clone $this;

        $duplicate->unload();

        $duplicate->setMulti($this->data);

        if ($duplicate->id_field) {
            $duplicate->setId($newId);
        }

        return $duplicate;
    }
```

Reply:
simply set entityId to null after clone, + then set whateer needed (unload, reset dirty, do your research)

Reply:
I can make it work but code is not consistent.
The problem I see is that clearing `entityId` should be done with `unload`. What is unload doing otherwise?

Reply:
Definitely not in unload, as this will open up entity which we do not want.

Reply:
But unload is exactly clearing the model from the entity.
I know it is "transitional" but something is not right.

Reply:
really?

Reply:
Is entityId cleared when cloning model? Probably should be.

Reply:
clone should not normally do these things

as a design rule, otherwise there is not a real issues

but to satisfy standards as well, we should do that in special method

now in duplciate() :)

Reply:
I am still of the opinion that if we support unload then entityId should be cleared. All other values are cleared.

Reply:
You got me for 1, 2 seconds. The answer is clear no.

> Entity approach must satisty several axioms. One of them is to never allow to assing different ID to the same instance.
> 
> So I think we should remove duplicate completely or clone before at the beginning of this method and return the cloned model then.

because if we allow to unset `entityId`, it can be easily changed, and this axiom will be violated. This any code that clears `entityId` is dangerous (except you know what you are dooing or do that on a fresh clone)

Reply:
Then we should remove `unload` method as it is not working properly at the moment or just use it to return a fresh object without entity loaded.

Reply:
`unload` unloads data, but keep entity bound to a specific ID

I think there are usecases for it.